Library/Homebrew/formulary.rb

```diff
@@ -902,47 +902,49 @@ def self.path(ref)
     loader_for(ref).path
   end
 
-  def self.tap_formula_name_path(tapped_name, warn:)
+  def self.tap_formula_name_type(tapped_name, warn:)
     user, repo, name = tapped_name.split("/", 3).map(&:downcase)
     tap = Tap.fetch user, repo
-    path = Formulary.find_formula_in_tap(name, tap)
-
-    unless path.file?
-      if (possible_alias = tap.alias_dir/name).file?
-        path = possible_alias.resolved_path
-        name = path.basename(".rb").to_s
-      elsif (new_name = tap.formula_renames[name].presence) &&
-            (new_path = Formulary.find_formula_in_tap(new_name, tap)).file?
-        old_name = name
-        path = new_path
-        name = new_name
-        new_name = tap.core_tap? ? name : "#{tap}/#{name}"
-      elsif (new_tap_name = tap.tap_migrations[name].presence)
-        new_tap_user, new_tap_repo, = new_tap_name.split("/")
-        new_tap_name = "#{new_tap_user}/#{new_tap_repo}"
-        new_tap = Tap.fetch new_tap_name
-        new_tap.ensure_installed!
-        new_tapped_name = "#{new_tap_name}/#{name}"
-        name, path, tap = Formulary.tap_formula_name_path(new_tapped_name, warn: false)
-        old_name = tapped_name
-        new_name = new_tap.core_tap? ? name : new_tapped_name
-      end
-
-      opoo "Formula #{old_name} was renamed to #{new_name}." if warn && old_name && new_name
+    type = nil
+
+    if (possible_alias = tap.alias_table[name].presence)
+      name = possible_alias
+      type = :alias
+    elsif (new_name = tap.formula_renames[name].presence)
+      old_name = name
+      name = new_name
+      new_name = tap.core_tap? ? name : "#{tap}/#{name}"
+      type = :rename
+    elsif (new_tap_name = tap.tap_migrations[name].presence)
+      new_tap_user, new_tap_repo, = new_tap_name.split("/")
+      new_tap_name = "#{new_tap_user}/#{new_tap_repo}"
+      new_tap = Tap.fetch new_tap_name
+      new_tap.ensure_installed!
+      new_tapped_name = "#{new_tap_name}/#{name}"
+      name, tap, = Formulary.tap_formula_name_type(new_tapped_name, warn: false)
+      old_name = tapped_name
+      new_name = new_tap.core_tap? ? name : new_tapped_name
+      type = :migration
```

Comment:
I guess the main difference in this logic at second glance is the removal of the `Formulary.find_formula_in_tap` and file path checks. Maybe those helped prevent infinite loops before and would have helped prevent what happened in https://github.com/Homebrew/brew/issues/16628.

CC: @dduugg

Reply:
Yes, previously it would just skip all this if the formula path exists.

Reply:
Thanks for the attention and fix here, @apainintheneck @reitermarkus 🥰

Reply:
@reitermarkus @apainintheneck @dduugg Should we revert this PR given https://github.com/Homebrew/brew/issues/16628 is still open?